server/src/Subscription.hs

```diff
@@ -64,22 +68,30 @@ unsubscribe path subid (SubscriptionTree here inner) =
 
 -- Invoke f for all subscribers to the path. The subscribers get passed the
 -- subvalue at the path that they are subscribed to.
-broadcast :: (Value -> v -> IO ()) -> [Text] -> Value -> SubscriptionTree k v -> IO ()
-broadcast f path value (SubscriptionTree here inner) =
+broadcast :: (v -> Value -> IO ()) -> [Text] -> Value -> SubscriptionTree k v -> IO ()
+broadcast f path value tree =
+  let notifications = execWriter $ broadcast' path value tree
+  in Async.mapConcurrently_ (uncurry f) notifications
+
+-- Like broadcast, but return a list of notifications rather than invoking an
+-- effect on each of them.
+broadcast' :: [Text] -> Value -> SubscriptionTree k v -> Writer [(v, Value)] ()
+broadcast' path value (SubscriptionTree here inner) = do
   case path of
     [] -> do
       -- When the path is empty, all subscribers that are "here" or at a deeper
       -- level should receive a notification.
       -- We broadcast concurrently since all updates are independent of each other
-      Async.forConcurrently_ here (f value)
-      let broadcastInner key = broadcast f [] (Store.lookupOrNull [key] value)
+      traverse_ (\v -> tell [(v, value)]) here
+      let broadcastInner key = broadcast' [] (Store.lookupOrNull [key] value)
       void $ HashMap.traverseWithKey broadcastInner inner
 
-    key : pathTail -> case HashMap.lookup key inner of
-      Nothing -> pure ()
-      -- TODO: Extract the inner thing from the value as well; the client is not
-      -- subscribed to the top-level thing after all.
-      Just subs -> broadcast f pathTail (Store.lookupOrNull [key] value) subs
+    key : pathTail -> do
+      traverse_ (\v -> tell [(v, value)]) here
+      for_ (HashMap.lookup key inner) $ \subs ->
+        -- TODO: Extract the inner thing from the value as well; the client is not
+        -- subscribed to the top-level thing after all.
```

Comment:
Isn't this fixed now? I thought that a client will now only receive the sub-tree that it is subscribed to and not the whole tree any more.

Reply:
Seems to have been already fixed here:

https://github.com/channable/icepeak/commit/77f8a90a59f616aa7b8631b109d4d57a866f3845

Will delete the comment.